pkg/services/baremetal/host/host.go

```diff
@@ -1072,6 +1073,42 @@ func validateStdOut(stdOut string) (string, error) {
 }
 
 // previous: Registering
+// next: ImageInstalling
+func (s *Service) actionPreProvisioning(ctx context.Context) actionResult {
+	markProvisionPending(s.scope.HetznerBareMetalHost, infrav1.StatePreProvisioning)
+	if s.scope.PreProvisionCommand == "" {
+		return actionComplete{}
+	}
+	creds := sshclient.CredentialsFromSecret(s.scope.RescueSSHSecret, s.scope.HetznerCluster.Spec.SSHKeys.RobotRescueSecretRef)
+	in := sshclient.Input{
+		PrivateKey: creds.PrivateKey,
+		Port:       rescuePort,
+		IP:         s.scope.HetznerBareMetalHost.Spec.Status.GetIPAddress(),
+	}
+	sshClient := s.scope.SSHClientFactory.NewClient(in)
+
+	// Ensure os ssh secret
+	sshKey, actResult := s.ensureSSHKey(s.scope.HetznerBareMetalHost.Spec.Status.SSHSpec.SecretRef, s.scope.OSSSHSecret)
```

Comment:
I think you should remove this code block from the actionImageInstalling. I think it has to be called only once

Reply:
ok, I did that. But I moved the code up, so that it is before "if ... return".